src/realm/node_header.hpp

```diff
@@ -229,16 +301,631 @@ class NodeHeader {
             case wtype_Ignore:
                 num_bytes = size;
                 break;
+            default: {
+                REALM_ASSERT(false);
+                break;
+            }
         }
-
+        num_bytes += header_size;
         // Ensure 8-byte alignment
         num_bytes = (num_bytes + 7) & ~size_t(7);
+        return num_bytes;
+    }
 
-        num_bytes += header_size;
+    // Possible header encodings (and corresponding memory layouts):
+    enum class Encoding { WTypBits, WTypMult, WTypIgn, Packed, AofP, PofA, Flex };
+    // ^ First 3 must overlap numerically with corresponding wtype_X enum.
+    static Encoding get_encoding(const char* header)
+    {
+        if (!wtype_is_extended(header)) {
+            return static_cast<Encoding>(get_wtype_from_header(header));
+        }
+        else {
+            const auto h = reinterpret_cast<const uint8_t*>(header);
+            return static_cast<Encoding>(h[5] + 3);
+        }
+    }
+    static void set_encoding(char* header, Encoding enc)
+    {
+        if (enc < Encoding::Packed) {
+            set_wtype_in_header(static_cast<WidthType>(enc), header);
+        }
+        else {
+            set_wtype_in_header(wtype_Extend, header);
+            auto h = reinterpret_cast<uint8_t*>(header);
+            h[5] = static_cast<uint8_t>(enc) - 3;
+        }
+    }
+    // * Packed: tightly packed array (any element size <= 64)
+    // * WTypBits: less tightly packed. Correspond to wtype_Bits
+    // * WTypMult: less tightly packed. Correspond to wtype_Multiply
+    // * WTypIgn: single byte elements. Correspond to wtype_Ignore
+    // encodings with more flexibility but lower number of elements:
+    // * AofP: Array of pairs (2 element sizes, 1 element count)
+    // * PofA: Pair of arrays (2 elememt sizes, 1 element count)
+    //   Choose between them according to spatial locality
+    // Encodings with even more flexibility with even lower number of elements
+    // * Flex: Pair of arrays (like PofA) but allowing different element count
+    //
+    // Encodings:     bytes:
+    // name:       |  b0   |  b1   |  b2   |  b3   | b4:0-2 | b4:3-4 | b4:5-7 |  b5   |  b6   |  b7  |
+    // oldies      |  cap/chksum           |  'A'  | width  | wtype  | flags  |          size        |
+    // Packed      |  cap/chksum   | -     | width | flags2 | wtype  | flags  | wtyp2 |     size     |
+    // AofP        |  cap/chksum   | w_A   | w_B   | flags2 | wtype  | flags  | wtyp2 |     size     |
+    // PofA        |  cap/chksum   | w_A   | w_B   | flags2 | wtype  | flags  | wtyp2 |     size     |
+    // Flex        |  cap/chksum   | w_B + size_B  | flags2 | wtype  | flags  | wtyp2 | w_A + size_A |
+    //
+    // legend: cap = capacity, chksum = checksum, flags = 3 flag bits, flags2 = 3 additional flag bits
```

Comment:
It is reserved to store the combination of encoding and type of array.
Like in this case for example (for packed):
`hb[5] = static_cast<uint8_t>(enc) - wtype_Extend;`

And it used mainly here:
```
static Encoding get_encoding(const char* header) {
   .....
   else {
            const auto h = reinterpret_cast<const uint8_t*>(header);
            return static_cast<Encoding>(h[5] + 3);
        }
}
```

```
enum class Encoding { WTypBits, WTypMult, WTypIgn, Packed, AofP, PofA, Flex };
enum WidthType {
        // The first 3 encodings where the only one used as far as Core v13.
        wtype_Bits = 0,     // width indicates how many bits every element occupies
        wtype_Multiply = 1, // width indicates how many bytes every element occupies
        wtype_Ignore = 2,   // each element is 1 byte
        wtype_Extend = 3    // the layouts are described in byte 4 of the header.  
    };
```
The reason for this is that, at the beginning, we simply stored BB in the header instead of AAAA (like we do today):
But we noticed a bunch of old formats did not have AAAA in the header, so in order to avoid rewriting the old array, during a commit or while compacting.  We re-arranged the new headers format.

Reply:
We want to save bits and have more flexibility in the future, in case we need to change stuff. We had a sparing bit inside WithType and this has saved a lot of rework, but we want to avoid being so tight. 
With type of array, we refer to the encoding as you pointed out, let me fix the comment.